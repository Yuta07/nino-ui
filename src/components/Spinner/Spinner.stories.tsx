import { storiesOf } from '@storybook/react';
import * as React from 'react';
import styled from 'styled-components';
// import components
import { Spinner } from './Spinner';

storiesOf('Spinner', module).add('all', () => (
  <React.Fragment>
    <Wrapper>
      <Spinner color="#c0392b" />
      <Spinner color="#f1c40f" />
      <Spinner color="#16a085" />
      <Spinner color="#2980b9" />
    </Wrapper>
  </React.Fragment>
));

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  padding: 1.5rem;
`;
