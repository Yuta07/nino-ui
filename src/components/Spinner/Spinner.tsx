import * as React from 'react';
import styled, { keyframes } from 'styled-components';

type Props = {
  color?: string;
  size?: 'small' | 'medium' | 'large';
};

export const Spinner = ({ color = '#f39c12' }: Props) => (
  <React.Fragment>
    <Loader color={color} />
  </React.Fragment>
);

// spinner style
// ref: [https://projects.lukehaas.me/css-loaders/]
const LoaderSpin = keyframes`
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
`;

const Loader = styled.div<{ color: string }>`
  margin: 0 auto;
  width: 3rem;
  height: 3rem;
  font-size: 5px;
  position: relative;
  border-radius: 50%;
  border-top: 0.3rem solid rgba(218, 218, 218, 0.2);
  border-right: 0.3rem solid rgba(218, 218, 218, 0.2);
  border-bottom: 0.3rem solid rgba(218, 218, 218, 0.2);
  border-left: 0.3rem solid ${props => props.color};
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
  -webkit-animation: ${LoaderSpin} 1.2s infinite linear;
  animation: ${LoaderSpin} 1.2s infinite linear;
`;
