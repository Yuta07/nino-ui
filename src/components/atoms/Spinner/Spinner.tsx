import * as React from 'react';
import styled, { css, keyframes } from 'styled-components';
import { Theme } from '../../../themes/Theme';
import { useTheme } from '../../../hooks/useTheme';

type Props = {
  color?: string;
  message?: string;
  position?: {
    top?: string;
    bottom?: string;
    left?: string;
    right?: string;
  };
};

export const Spinner = ({ color = 'MAIN' }: Props) => {
  const themes = useTheme();

  return <Loader color={color} themes={themes} />;
};

export const SpinnerWithMessage = ({ color = 'MAIN', message, position }: Props) => {
  const themes = useTheme();

  return (
    <Container>
      <Loader color={color} themes={themes} />
      <Message themes={themes} posi={position}>
        {message}
      </Message>
    </Container>
  );
};

// ref: [https://projects.lukehaas.me/css-loaders/]

const Container = styled.div`
  position: relative;
`;

const LoaderSpin = keyframes`
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
`;

const Loader = styled.div<{ color: Props['color']; themes: Theme }>`
  ${({ color, themes }) => {
    const { device, palette } = themes;

    return css`
      width: 2.5rem;
      height: 2.5rem;
      font-size: 5px;
      border-radius: 50%;
      border-top: 0.3rem solid rgba(218, 218, 218, 0.2);
      border-right: 0.3rem solid rgba(218, 218, 218, 0.2);
      border-bottom: 0.3rem solid rgba(218, 218, 218, 0.2);
      border-left: 0.3rem solid ${palette[color]};
      -webkit-transform: translateZ(0);
      -ms-transform: translateZ(0);
      transform: translateZ(0);
      -webkit-animation: ${LoaderSpin} 1.2s infinite linear;
      animation: ${LoaderSpin} 1.2s infinite linear;

      @media screen and ${device.TABLET} {
        width: 2rem;
        height: 2rem;
        border-top: 0.25rem solid rgba(218, 218, 218, 0.2);
        border-right: 0.25rem solid rgba(218, 218, 218, 0.2);
        border-bottom: 0.25rem solid rgba(218, 218, 218, 0.2);
        border-left: 0.25rem solid ${palette[color]};
      }

      @media screen and ${device.MOBILE} {
        width: 1.5rem;
        height: 1.5rem;
        border-top: 0.2rem solid rgba(218, 218, 218, 0.2);
        border-right: 0.2rem solid rgba(218, 218, 218, 0.2);
        border-bottom: 0.2rem solid rgba(218, 218, 218, 0.2);
        border-left: 0.2rem solid ${palette[color]};
      }
    `;
  }}
`;

const Message = styled.span<{ themes: Theme; posi: Props['position'] }>`
  ${({ themes, posi }) => {
    const { device } = themes;
    const { top, bottom, left, right } = posi;

    return css`
      position: absolute;
      font-size: 14px;
      white-space: nowrap;

      ${top &&
        css`
          top: ${top};
        `}
      ${bottom &&
        css`
          bottom: ${bottom};
        `}
      ${right &&
        css`
          right: ${right};
        `}
      ${left &&
        css`
          left: ${left};
        `}

      @media screen and ${device.MOBILE} {
        font-size: 12px;
      }
    `;
  }}
`;
