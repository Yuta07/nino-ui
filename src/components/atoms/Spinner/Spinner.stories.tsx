import { storiesOf } from '@storybook/react';
import * as React from 'react';
import styled from 'styled-components';
import { Spinner, SpinnerWithMessage } from './Spinner';

storiesOf('Spinner', module).add('all', () => (
  <Wrapper>
    <Container>
      <Spinner color="MAIN" />
    </Container>
    <Container>
      <Spinner color="DANGER" />
    </Container>
    <Container>
      <Spinner color="SUCCESS" />
    </Container>
    <Container>
      <SpinnerWithMessage color="WARNING" message="Now Loading..." position={{ top: '50px', left: '-100%' }} />
    </Container>
    <Container>
      <SpinnerWithMessage color="INFO" message="読み込み中..." position={{ top: '50px', left: '-50%' }} />
    </Container>
  </Wrapper>
));

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-around;
  padding: 1.5rem;
`;

const Container = styled.div``;
