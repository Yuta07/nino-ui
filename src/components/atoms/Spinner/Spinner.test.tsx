import React from 'react';
import * as renderer from 'react-test-renderer';

import { Spinner } from '../Spinner';

describe('Spinner test', function() {
  test('Spinner snapshot', () => {
    const tree = renderer.create(<Spinner />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
