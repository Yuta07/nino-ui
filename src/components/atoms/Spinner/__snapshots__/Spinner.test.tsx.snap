// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner test Spinner snapshot 1`] = `
.c0 {
  margin: 0 auto;
  width: 3rem;
  height: 3rem;
  font-size: 5px;
  position: relative;
  border-radius: 50%;
  border-top: 0.3rem solid rgba(218,218,218,0.2);
  border-right: 0.3rem solid rgba(218,218,218,0.2);
  border-bottom: 0.3rem solid rgba(218,218,218,0.2);
  border-left: 0.3rem solid #f39c12;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
  -webkit-animation: cilQsd 1.2s infinite linear;
  -webkit-animation: cilQsd 1.2s infinite linear;
  animation: cilQsd 1.2s infinite linear;
}

@media screen and (max-width:768px) {
  .c0 {
    width: 2.5rem;
    height: 2.5rem;
    border-top: 0.25rem solid rgba(218,218,218,0.2);
    border-right: 0.25rem solid rgba(218,218,218,0.2);
    border-bottom: 0.25rem solid rgba(218,218,218,0.2);
    border-left: 0.25rem solid #f39c12;
  }
}

@media screen and (max-width:480px) {
  .c0 {
    width: 2rem;
    height: 2rem;
    border-top: 0.2rem solid rgba(218,218,218,0.2);
    border-right: 0.2rem solid rgba(218,218,218,0.2);
    border-bottom: 0.2rem solid rgba(218,218,218,0.2);
    border-left: 0.2rem solid #f39c12;
  }
}

<div
  className="c0"
  color="MAIN"
/>
`;
